File: algorithms/ruby/2187-minimum-time-to-complete-trips.rb
# frozen_string_literal: true

# Problem: 2187. Minimum Time to Complete Trips
# URL: https://leetcode.com/problems/minimum-time-to-complete-trips

# @param {Integer[]} time
# @param {Integer} total_trips
# @return {Integer}
def minimum_time(time, total_trips)
  (1..time.min * total_trips).bsearch do |x|
    time.sum { x / _1 } >= total_trips
  end
end

# **************** #
#       TEST       #
# **************** #

require "test/unit"
class Test_minimum_time < Test::Unit::TestCase
  def test_
    assert_equal 3, minimum_time([1, 2, 3], 5)
    assert_equal 2, minimum_time([2], 1)
    assert_equal 4, minimum_time([1, 2, 3, 1], 10)
  end
end

File: algorithms/python3/2187-minimum-time-to-complete-trips.py
# 2187. Minimum Time to Complete Trips
# https://leetcode.com/problems/minimum-time-to-complete-trips

class Solution:
	def minimumTime(self, time: list[int], totalTrips: int) -> int:
		# Initialize the left and right boundaries.
		left, right = 1, min(time) * totalTrips

		# Can these buses finish 'totalTrips' of trips in 'given_time'?
		def timeEnough(given_time):
			return sum(given_time // t for t in time) >= totalTrips

		# Binary search to find the minimum time to finish the task.
		while left < right:
			mid = (left + right) // 2
			if timeEnough(mid):
				right = mid
			else:
				left = mid + 1
		return left

# ********************#
#       TEST         #
# ********************#

import unittest

class TestMinimumTime(unittest.TestCase):
  def test_minimumTime(self):
    self.assertEqual(Solution.minimumTime(self, [1, 2, 3, 5], 5), 3)
    self.assertEqual(Solution.minimumTime(self, [2], 1),2)

if __name__ == '__main__':
  unittest.main()
